src/catalua.cpp

```diff
@@ -77,47 +92,279 @@ bool lua_report_error(lua_State *L, int err, const char *path) {
+        lua_pushvalue( L, -1 );
+        // Set the metatable of the new metatable (-2 on the stack) to be itself. Pretty meta, hu?
+        // It also pops one value from the stack.
+        lua_setmetatable( L, -2 );
+        // Now set the actual functions of the metatable.
+        luaL_setfuncs( L, &FUNCTIONS[0], 0 );
+
+        // Calling the destructor is really only needed when it's not trivial (e.g. pointers)
+        if( !std::is_trivially_destructible<T>::value ) {
+            // Push function pointer
+            lua_pushcfunction( L, &gc );
+            // -1 would be the function pointer, -2 is the metatable, the function pointer is popped
+            lua_setfield( L, -2, "__gc" );
+        }
+        lua_pushcfunction( L, &index );
+        lua_setfield( L, -2, "__index" );
+        lua_pushcfunction( L, &newindex );
+        lua_setfield( L, -2, "__newindex" );
+    }
 
-        luah_setglobal(L, "player", -1);
+public:
+    static void load_metatable( lua_State* const L )
+    {
+        get_metatable( L );
+        std::string mt(METATABLE_NAME);
+        if( mt.length() > 10 ) {
+            // 10 == strlen("_metatable"), the prefix that all metatables have, but we want them to
+            // be available without this in the global namespace (e.g. to write `p = tripoint( 1, 5, 100 )`
+            mt.erase( mt.length() - 10, 10 );
+        }
+        lua_setglobal( L, mt.c_str() );
+    }
+    static void push( lua_State* const L, const T& value )
+    {
+        // Push user data,
+        T* value_in_lua = static_cast<T*>( lua_newuserdata( L, sizeof( T ) ) );
+        // Push metatable,
+        get_metatable( L );
+        // -1 would the the metatable, -2 is the uservalue, the table is popped
+        lua_setmetatable( L, -2 );
+        // This is where the copy happens:
+        new (value_in_lua) T( value );
+    }
+    static int push_reg( lua_State* const L, const T& value )
+    {
+        push( L, value );
+        return luah_store_in_registry( L, -1 );
+    }
+    static T& get( lua_State* const L, int const stack_index )
+    {
+        luaL_checktype( L, stack_index, LUA_TUSERDATA );
+        T* user_data = static_cast<T*>( lua_touserdata( L, stack_index ) );
+        if( user_data == nullptr ) {
+            // luaL_error does not return at all.
+            return *(T*) luaL_error( L, "First argument to function is not a class" );
+        }
+        return *user_data;
+    }
+    /** Compatibility with the function in @ref LuaReference. But this *always* returns a reference. */
+    static T &get_proxy( lua_State * const L, int const stack_position )
+    {
+        return get( L, stack_position );
     }
+};
 
-    // Make sure the map reference is up to date.
+/**
+ * This is special wrapper (an extension) for references to objects which are not stored in Lua,
+ * but are kept in the memory managed by C++. This class only stored are retrieves the pointers,
```

Comment:
So this class is supposed to do one thing less than its super class: It's not supposed to call the destructor on the value when the lua value is garbage collected.